firedrake/slate/slac/compiler.py

```diff
@@ -530,6 +590,325 @@ def parenthesize(arg, prec=None, parent=None):
+
+        def with_types(self, arg_id_to_dtype, kernel, callables_table):
+            for i in range(0, len(arg_id_to_dtype)):
+                if i not in arg_id_to_dtype or arg_id_to_dtype[i] is None:
+                    # the types provided aren't mature enough to specialize the
+                    # callable
+                    return (self.copy(arg_id_to_dtype=arg_id_to_dtype),
+                            callables_table)
+
+            mat_dtype = arg_id_to_dtype[0].numpy_dtype
+            name_in_target = "inverse_"
+
+            from loopy.types import NumpyType
+            return (self.copy(name_in_target=name_in_target,
+                              arg_id_to_dtype={0: NumpyType(mat_dtype), 1: NumpyType(int)}),
+                    callables_table)
+
+        def emit_call_insn(self, insn, target, expression_to_code_mapper):
+            assert self.is_ready_for_codegen()
+
+            assert isinstance(insn, loopy.CallInstruction)
+
+            parameters = insn.expression.parameters
+
+            parameters = list(parameters)
+            par_dtypes = [self.arg_id_to_dtype[i] for i, _ in enumerate(parameters)]
+
+            parameters.append(insn.assignees[0])
+            par_dtypes.append(self.arg_id_to_dtype[0])
+
+            from loopy.expression import dtype_to_type_context
+            from pymbolic.mapper.stringifier import PREC_NONE
+            from loopy.symbolic import SubArrayRef
+            from pymbolic import var
+
+            mat_descr = self.arg_id_to_descr[0]
+
+            arg_c_parameters = [
+                expression_to_code_mapper(
+                    par,
+                    PREC_NONE,
+                    dtype_to_type_context(target, par_dtype),
+                    par_dtype
+                ).expr
+                if isinstance(par, SubArrayRef) else
+                expression_to_code_mapper(
+                    par,
+                    PREC_NONE,
+                    dtype_to_type_context(target, par_dtype),
+                    par_dtype
+                ).expr
+                for par, par_dtype in zip(parameters, par_dtypes)
+            ]
+            c_parameters = []
+            c_parameters.insert(0, arg_c_parameters[0])  # t1
+            c_parameters.insert(1, mat_descr.shape[0])  # n
+            return var(self.name_in_target)(*c_parameters), False
+
+        def generate_preambles(self, target):
+            assert isinstance(target, CTarget)
+            inverse_preamble = """
+                #include <string.h>
+                #include <stdio.h>
+                #include <stdlib.h>
+                #ifndef Inverse_HPP
+                #define Inverse_HPP
+                void inverse_(PetscScalar* A, PetscBLASInt N)
+                {
+                    PetscBLASInt info;
+                    PetscBLASInt* ipiv=(PetscBLASInt*) malloc(N*sizeof(PetscBLASInt));
+                    PetscScalar* Awork=(PetscScalar*) malloc(N*N*sizeof(PetscScalar));
```

Comment:
So who is going to free these memory blocks?

Reply:
I am afraid I don't understand/know. The C compiler I guess?

Reply:
I should probably free the memory as well, however.

Reply:
It's probably a memory leak then, so you should take care the memory released.